src/core/lib/event_engine/poller.h

```diff
@@ -0,0 +1,53 @@
+// Copyright 2022 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_CORE_LIB_EVENT_ENGINE_POLLER_H
+#define GRPC_CORE_LIB_EVENT_ENGINE_POLLER_H
+
+#include <grpc/support/port_platform.h>
+
+#include "absl/container/inlined_vector.h"
+#include "absl/status/statusor.h"
+
+#include <grpc/event_engine/event_engine.h>
+
+namespace grpc_event_engine {
+namespace experimental {
+
+namespace {
+constexpr size_t kInitialSize = 5;
+}  // namespace
+
+// A generic cross-platform "poller" concept.
+// Concrete implementations will likely manage a set of sockets/file
+// descriptors/etc, allowing threads to drive polling and event processing via
+// Work(...).
+class Poller {
+ public:
+  using Events = absl::InlinedVector<EventEngine::Closure*, kInitialSize>;
+  virtual ~Poller() = default;
+  // Poll once for events, returning a collection of Closures to be executed.
+  //
+  // Returns:
+  //  * absl::AbortedError if it was Kicked.
+  //  * absl::DeadlineExceeded if timeout occurred
+  //  * A collection of closures to execute, otherwise
+  virtual absl::StatusOr<Events> Work(EventEngine::Duration timeout) = 0;
```

Comment:
AbortedError and DeadlineExceeded both require a string which must be copied - and consequently an allocation made to hold that string. Since this is fast path and both kicks and deadline exceeded are on that fast path, let's instead define this as:

```
using Events = ...;
struct DeadlineExceeded {};
struct Kicked {};
using WorkResult = absl::variant<Events, DeadlineExceeded, Kicked>;
virtual WorkResult Work(Duration timeout) = 0;
```

Reply:
Done